File: backend/app/api/alerts.py
"""
Tesis → alertas automáticas — el diferencial de FinPulse.

La IA lee las tesis del usuario (Decision Journal, documentos, o texto manual),
extrae las condiciones que las INVALIDARÍAN (niveles de precio) y las vigila
contra precios reales. Cuando un nivel se cruza, la alerta se dispara: la tesis
está tocada o rota, y el usuario se entera antes de racionalizar la pérdida.

Extracción: Claude si hay API key (entiende la tesis y propone niveles con
racional); fallback heurístico por regex de niveles si no.
Chequeo: POST /alerts/check evalúa todas las activas con yfinance — lo llamará
un cron; mientras tanto el frontend lo invoca al cargar (mismo patrón que el
recalc del tracking).
"""

import asyncio
import json
import re
from datetime import datetime
from typing import Optional

import anthropic
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.database import get_db
from app.models.models import Decision, ThesisAlert
from app.api.auth import get_current_user, User
from app.services.market import get_quotes

router = APIRouter(prefix="/alerts", tags=["alerts"])


# ── Schemas ──

class Proposal(BaseModel):
    condition: str  # price_below / price_above
    level: float
    severity: str  # aviso / invalidacion
    rationale: str = ""


class ExtractRequest(BaseModel):
    ticker: str
    thesis: str
    action: str = "buy"  # buy / sell — orienta la dirección de invalidación


class ExtractResponse(BaseModel):
    ticker: str
    current_price: Optional[float]
    thesis_summary: str
    proposals: list[Proposal]
    engine: str  # ia / heuristica


class AlertCreate(BaseModel):
    ticker: str
    thesis_summary: str
    condition: str
    level: float
    severity: str = "invalidacion"
    rationale: str = ""
    source_type: str = "manual"  # journal / document / manual
    source_id: Optional[int] = None


class AlertOut(BaseModel):
    id: int
    ticker: str
    thesis_summary: str
    source_type: str
    source_id: Optional[int]
    condition: str
    level: float
    severity: str
    rationale: str
    status: str
    triggered_at: Optional[datetime]
    triggered_price: Optional[float]
    last_price: Optional[float]
    last_checked_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


class AlertPatch(BaseModel):
    status: str  # active / dismissed


class CheckResult(BaseModel):
    checked: int
    triggered: list[AlertOut]
    alerts: list[AlertOut]


class ScanResult(BaseModel):
    scanned_theses: int
    created: list[AlertOut]
    engine: str
    message: str


# ── Extracción de condiciones de invalidación ──

EXTRACT_PROMPT = """Eres el CIO de FinPulse. Tu tarea: leer la tesis de inversión del usuario y extraer las condiciones de PRECIO que la INVALIDARÍAN.

Reglas:
- Solo niveles de precio del activo de la tesis (no de otros activos).
- Si la tesis menciona niveles explícitos (soportes, stops, "si cae por debajo de X"), úsalos.
- Si no hay niveles explícitos pero la tesis es direccional, propón niveles razonables: para una tesis alcista, un aviso ~-8% y una invalidación ~-15% desde el precio actual; para una bajista, al revés.
- Máximo 3 condiciones. severity: "aviso" (la tesis se debilita) o "invalidacion" (la tesis está rota).
- rationale: UNA frase concreta conectando el nivel con la tesis.
- Si la tesis no da pie a ninguna condición de precio, devuelve [].

Responde SOLO con JSON válido, sin texto adicional ni markdown:
{"thesis_summary": "<resumen de la tesis en 1 frase>", "conditions": [{"condition": "price_below"|"price_above", "level": <número>, "severity": "aviso"|"invalidacion", "rationale": "<frase>"}]}"""


def _extract_levels(text: str) -> list[float]:
    """Encuentra niveles de precio mencionados en el texto ($66, 66,40€, 'los 62'...)."""
    levels = []
    for m in re.finditer(r"[$€£]\s?(\d+(?:[.,]\d+)?)|(\d+(?:[.,]\d+)?)\s?[$€£]|\blos\s+(\d+(?:[.,]\d+)?)\b", text):
        raw = next(g for g in m.groups() if g)
        try:
            levels.append(float(raw.replace(",", ".")))
        except ValueError:
            continue
    return levels


def _heuristic_extract(ticker: str, thesis: str, action: str, current_price: float | None) -> tuple[str, list[Proposal]]:
    """Sin API key: regex de niveles + dirección según la acción de la tesis."""
    summary = thesis.strip().replace("\n", " ")
    if len(summary) > 180:
        summary = summary[:177] + "..."

    levels = _extract_levels(thesis)
    bullish = action != "sell"
    condition = "price_below" if bullish else "price_above"

    # Niveles relevantes: los que invalidarían (por debajo si alcista, por encima si bajista)
    if current_price:
        relevant = sorted(
            [lv for lv in levels if (lv < current_price) == bullish or lv == current_price],
            reverse=bullish,
        )
    else:
        relevant = sorted(set(levels), reverse=bullish)

    proposals: list[Proposal] = []
    if relevant:
        capped = relevant[:2] if len(relevant) >= 2 else relevant
        if len(capped) == 2:
            proposals.append(Proposal(condition=condition, level=capped[0], severity="aviso",
                                      rationale=f"Primer nivel mencionado en la tesis: la tesis se debilita en {capped[0]:g}."))
            proposals.append(Proposal(condition=condition, level=capped[1], severity="invalidacion",
                                      rationale=f"Nivel de ruptura de la tesis: {capped[1]:g}."))
        else:
            proposals.append(Proposal(condition=condition, level=capped[0], severity="invalidacion",
                                      rationale=f"Único nivel citado en la tesis: {capped[0]:g}."))
    elif current_price:
        aviso = round(current_price * (0.92 if bullish else 1.08), 2)
        invalida = round(current_price * (0.85 if bullish else 1.15), 2)
        proposals.append(Proposal(condition=condition, level=aviso, severity="aviso",
                                  rationale="Sin niveles explícitos en la tesis: aviso a ±8% del precio actual."))
        proposals.append(Proposal(condition=condition, level=invalida, severity="invalidacion",
                                  rationale="Invalidación por defecto a ±15% del precio actual."))

    return summary, proposals


def _parse_claude_json(text: str) -> dict:
    cleaned = text.strip()
    cleaned = re.sub(r"^```(?:json)?\s*|\s*```$", "", cleaned, flags=re.MULTILINE).strip()
    start = cleaned.find("{")
    end = cleaned.rfind("}")
    if start == -1 or end == -1:
        raise ValueError("Sin JSON en la respuesta")
    return json.loads(cleaned[start : end + 1])


async def _ai_extract(ticker: str, thesis: str, action: str, current_price: float | None) -> tuple[str, list[Proposal]] | None:
    """Claude lee la tesis y propone condiciones. None si no hay API key o falla."""
    if not settings.anthropic_api_key:
        return None
    context = f"""Activo: {ticker}
Acción de la tesis: {"COMPRA (tesis alcista)" if action != "sell" else "VENTA (tesis bajista)"}
Precio actual: {f"{current_price:.2f}" if current_price else "desconocido"}

TESIS DEL USUARIO:
{thesis}"""
    try:
        client = anthropic.Anthropic(api_key=settings.anthropic_api_key)
        response = await asyncio.to_thread(
            lambda: client.messages.create(
                model="claude-sonnet-4-20250514",
                max_tokens=1024,
                system=EXTRACT_PROMPT,
                messages=[{"role": "user", "content": context}],
            )
        )
        data = _parse_claude_json(response.content[0].text)
        proposals = []
        for c in data.get("conditions", [])[:3]:
            if c.get("condition") in ("price_below", "price_above") and isinstance(c.get("level"), (int, float)):
                proposals.append(Proposal(
                    condition=c["condition"],
                    level=float(c["level"]),
                    severity=c.get("severity") if c.get("severity") in ("aviso", "invalidacion") else "invalidacion",
                    rationale=str(c.get("rationale", ""))[:500],
                ))
        summary = str(data.get("thesis_summary", ""))[:300] or thesis[:180]
        return summary, proposals
    except Exception:
        return None


async def _current_price(ticker: str) -> float | None:
    try:
        quotes = await get_quotes([ticker])
        price = quotes.get(ticker.upper(), {}).get("price") or quotes.get(ticker, {}).get("price")
        return float(price) if price else None
    except Exception:
        return None


async def _extract(ticker: str, thesis: str, action: str) -> ExtractResponse:
    price = await _current_price(ticker)
    ai = await _ai_extract(ticker, thesis, action, price)
    if ai is not None:
        summary, proposals = ai
        engine = "ia"
    else:
        summary, proposals = _heuristic_extract(ticker, thesis, action, price)
        engine = "heuristica"
    return ExtractResponse(ticker=ticker.upper(), current_price=price, thesis_summary=summary,
                           proposals=proposals, engine=engine)


# ── Endpoints ──

@router.post("/extract", response_model=ExtractResponse)
async def extract_conditions(data: ExtractRequest, user: User = Depends(get_current_user)):
    """La IA lee una tesis y propone las alertas de invalidación (sin persistir)."""
    if not data.thesis.strip():
        raise HTTPException(status_code=400, detail="La tesis no puede estar vacía")
    return await _extract(data.ticker, data.thesis, data.action)


@router.post("", response_model=list[AlertOut])
async def create_alerts(
    alerts: list[AlertCreate],
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Persiste alertas (normalmente las propuestas por /extract que el usuario acepta)."""
    if not alerts:
        raise HTTPException(status_code=400, detail="Lista de alertas vacía")
    created = []
    for a in alerts:
        if a.condition not in ("price_below", "price_above"):
            raise HTTPException(status_code=400, detail=f"Condición inválida: {a.condition}")
        alert = ThesisAlert(
            user_id=user.id,
            ticker=a.ticker.upper(),
            thesis_summary=a.thesis_summary,
            source_type=a.source_type,
            source_id=a.source_id,
            condition=a.condition,
            level=a.level,
            severity=a.severity if a.severity in ("aviso", "invalidacion") else "invalidacion",
            rationale=a.rationale,
        )
        db.add(alert)
        created.append(alert)
    await db.commit()
    for alert in created:
        await db.refresh(alert)
    return [AlertOut.model_validate(a) for a in created]


@router.get("", response_model=list[AlertOut])
async def list_alerts(
    status: str = "all",  # all / active / triggered / dismissed
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    q = select(ThesisAlert).where(ThesisAlert.user_id == user.id)
    if status != "all":
        q = q.where(ThesisAlert.status == status)
    result = await db.execute(q.order_by(ThesisAlert.created_at.desc()).limit(100))
    return [AlertOut.model_validate(a) for a in result.scalars().all()]


@router.patch("/{alert_id}", response_model=AlertOut)
async def patch_alert(
    alert_id: int,
    data: AlertPatch,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Descartar o reactivar una alerta."""
    if data.status not in ("active", "dismissed"):
        raise HTTPException(status_code=400, detail="status debe ser active o dismissed")
    result = await db.execute(
        select(ThesisAlert).where(ThesisAlert.id == alert_id, ThesisAlert.user_id == user.id)
    )
    alert = result.scalar_one_or_none()
    if not alert:
        raise HTTPException(status_code=404, detail="Alerta no encontrada")
    alert.status = data.status
    if data.status == "active":
        alert.triggered_at = None
        alert.triggered_price = None
    await db.commit()
    await db.refresh(alert)
    return AlertOut.model_validate(alert)


@router.post("/check", response_model=CheckResult)
async def check_alerts(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    El vigilante: evalúa todas las alertas activas contra precios reales.
    Lo llamará un cron; mientras tanto, el frontend lo invoca al cargar.
    """
    result = await db.execute(
        select(ThesisAlert).where(ThesisAlert.user_id == user.id, ThesisAlert.status == "active")
    )
    active = list(result.scalars().all())
    if not active:
        return CheckResult(checked=0, triggered=[], alerts=[])

    tickers = sorted({a.ticker for a in active})
    try:
        quotes = await get_quotes(tickers)
    except Exception:
        quotes = {}

    now = datetime.utcnow()
    triggered = []
    for a in active:
        price = quotes.get(a.ticker, {}).get("price")
        if not price:
            continue
        a.last_price = float(price)
        a.last_checked_at = now
        fired = (a.condition == "price_below" and price <= a.level) or (
            a.condition == "price_above" and price >= a.level
        )
        if fired:
            a.status = "triggered"
            a.triggered_at = now
            a.triggered_price = float(price)
            triggered.append(a)

    await db.commit()
    for a in active:
        await db.refresh(a)

    return CheckResult(
        checked=len(active),
        triggered=[AlertOut.model_validate(a) for a in triggered],
        alerts=[AlertOut.model_validate(a) for a in active],
    )


@router.post("/scan-journal", response_model=ScanResult)
async def scan_journal(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Recorre las tesis escritas en el Decision Journal y crea alertas de
    invalidación para las que aún no tienen ninguna. La IA (o la heurística)
    lee cada tesis una sola vez.
    """
    result = await db.execute(
        select(Decision).where(Decision.user_id == user.id, Decision.thesis != "")
    )
    decisions = [d for d in result.scalars().all() if (d.thesis or "").strip()]

    existing = await db.execute(
        select(ThesisAlert.source_id).where(
            ThesisAlert.user_id == user.id, ThesisAlert.source_type == "journal"
        )
    )
    covered = {row[0] for row in existing.all()}

    to_scan = [d for d in decisions if d.id not in covered]
    created: list[ThesisAlert] = []
    engine = "heuristica"

    for d in to_scan:
        extraction = await _extract(d.ticker, d.thesis, d.action)
        if extraction.engine == "ia":
            engine = "ia"
        for p in extraction.proposals:
            alert = ThesisAlert(
                user_id=user.id,
                ticker=d.ticker,
                thesis_summary=extraction.thesis_summary,
                source_type="journal",
                source_id=d.id,
                condition=p.condition,
                level=p.level,
                severity=p.severity,
                rationale=p.rationale,
            )
            db.add(alert)
            created.append(alert)

    await db.commit()
    for a in created:
        await db.refresh(a)

    return ScanResult(
        scanned_theses=len(to_scan),
        created=[AlertOut.model_validate(a) for a in created],
        engine=engine,
        message=f"{len(created)} alertas creadas de {len(to_scan)} tesis nuevas"
        if to_scan else "Todas las tesis del journal ya están vigiladas",
    )

File: frontend/src/lib/alerts.ts
/**
 * Tesis → alertas automáticas — cliente API + fallback demo.
 * La IA lee las tesis (Decision Journal, documentos o texto manual), extrae
 * los niveles de precio que las invalidarían y los vigila con precios reales
 * (POST /api/alerts/check). Sin sesión/backend: modo demo coherente con el
 * journal, con persistencia local de descartes y altas.
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";
const LOCAL_KEY = "finpulse-alerts-local-v1";

// ── Types ──

export type AlertCondition = "price_below" | "price_above";
export type AlertSeverity = "aviso" | "invalidacion";
export type AlertStatus = "active" | "triggered" | "dismissed";

export type ThesisAlert = {
  id: number;
  ticker: string;
  thesis_summary: string;
  source_type: "journal" | "document" | "manual";
  source_id: number | null;
  condition: AlertCondition;
  level: number;
  severity: AlertSeverity;
  rationale: string;
  status: AlertStatus;
  triggered_at: string | null;
  triggered_price: number | null;
  last_price: number | null;
  last_checked_at: string | null;
  created_at: string;
};

export type Proposal = {
  condition: AlertCondition;
  level: number;
  severity: AlertSeverity;
  rationale: string;
};

export type ExtractResult = {
  ticker: string;
  current_price: number | null;
  thesis_summary: string;
  proposals: Proposal[];
  engine: "ia" | "heuristica";
};

export type ScanResult = {
  scanned_theses: number;
  created: ThesisAlert[];
  engine: string;
  message: string;
};

export type AlertsData = {
  alerts: ThesisAlert[];
  demo: boolean;
};

// ── Auth + fetch ──

function authHeaders(): Record<string, string> {
  if (typeof window === "undefined") return {};
  const token = localStorage.getItem("finpulse_token");
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function api<T>(path: string, options: RequestInit = {}): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    ...options,
    headers: { ...authHeaders(), ...options.headers },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || `API error ${res.status}`);
  }
  return res.json();
}

// ── Mock (coherente con el Decision Journal demo) ──

export const MOCK_ALERTS: ThesisAlert[] = [
  // Tesis manual del usuario: Brent con niveles explícitos $66/$62
  {
    id: 501, ticker: "BRT",
    thesis_summary: "La prima geopolítica sostiene el Brent por encima de los 66; por debajo de 62 la tesis está rota.",
    source_type: "manual", source_id: null,
    condition: "price_below", level: 66, severity: "aviso",
    rationale: "Primer nivel citado en la tesis: perder los 66 debilita el argumento de la prima geopolítica.",
    status: "triggered", triggered_at: "2026-07-02T15:30:00", triggered_price: 65.82,
    last_price: 64.9, last_checked_at: "2026-07-07T08:00:00", created_at: "2026-06-20T10:00:00",
  },
  {
    id: 502, ticker: "BRT",
    thesis_summary: "La prima geopolítica sostiene el Brent por encima de los 66; por debajo de 62 la tesis está rota.",
    source_type: "manual", source_id: null,
    condition: "price_below", level: 62, severity: "invalidacion",
    rationale: "Nivel de ruptura: bajo 62 la oferta OPEC+ domina y la tesis queda invalidada.",
    status: "active", triggered_at: null, triggered_price: null,
    last_price: 64.9, last_checked_at: "2026-07-07T08:00:00", created_at: "2026-06-20T10:00:00",
  },
  // Extraídas del journal (decisión SEMI, tesis capex IA — sin niveles explícitos → ±8/15%)
  {
    id: 503, ticker: "SEMI",
    thesis_summary: "El ciclo de capex en IA sigue acelerando; SEMI como exposición diversificada al sector.",
    source_type: "journal", source_id: 101,
    condition: "price_below", level: 40.5, severity: "aviso",
    rationale: "Sin niveles explícitos en la tesis: aviso a -8% del precio de entrada al ciclo.",
    status: "active", triggered_at: null, triggered_price: null,
    last_price: 45.93, last_checked_at: "2026-07-07T08:00:00", created_at: "2026-07-01T09:00:00",
  },
  {
    id: 504, ticker: "SEMI",
    thesis_summary: "El ciclo de capex en IA sigue acelerando; SEMI como exposición diversificada al sector.",
    source_type: "journal", source_id: 101,
    condition: "price_below", level: 37.4, severity: "invalidacion",
    rationale: "Una caída del 15% implicaría que el mercado descuenta el fin del ciclo de capex.",
    status: "active", triggered_at: null, triggered_price: null,
    last_price: 45.93, last_checked_at: "2026-07-07T08:00:00", created_at: "2026-07-01T09:00:00",
  },
  // Extraída del journal (decisión EUNA, tesis recorte BCE con dato Polymarket)
  {
    id: 505, ticker: "EUNA",
    thesis_summary: "Polymarket da 78% a recorte del BCE en septiembre: alargar duración antes del movimiento.",
    source_type: "journal", source_id: 105,
    condition: "price_below", level: 4.44, severity: "aviso",
    rationale: "Si el bono cae un 4%, el mercado está descontando que el recorte no llega.",
    status: "active", triggered_at: null, triggered_price: null,
    last_price: 4.66, last_checked_at: "2026-07-07T08:00:00", created_at: "2026-07-02T11:10:00",
  },
];

// ── Persistencia local (modo demo) ──

type LocalState = { created: ThesisAlert[]; dismissed: number[] };

function readLocal(): LocalState {
  if (typeof window === "undefined") return { created: [], dismissed: [] };
  try {
    const raw = localStorage.getItem(LOCAL_KEY);
    if (raw) return JSON.parse(raw);
  } catch {}
  return { created: [], dismissed: [] };
}

function writeLocal(state: LocalState) {
  try {
    localStorage.setItem(LOCAL_KEY, JSON.stringify(state));
  } catch {}
}

function demoAlerts(): ThesisAlert[] {
  const local = readLocal();
  return [...local.created, ...MOCK_ALERTS].map((a) =>
    local.dismissed.includes(a.id) ? { ...a, status: "dismissed" as AlertStatus } : a
  );
}

// ── API pública (real con fallback demo) ──

export async function loadAlerts(): Promise<AlertsData> {
  try {
    // check refresca precios y dispara lo que toque; luego traemos el estado completo
    await api("/api/alerts/check", { method: "POST" }).catch(() => null);
    const alerts = await api<ThesisAlert[]>("/api/alerts?status=all");
    return { alerts, demo: false };
  } catch {
    return { alerts: demoAlerts(), demo: true };
  }
}

export async function extractFromThesis(ticker: string, thesis: string, action: string): Promise<ExtractResult> {
  return api<ExtractResult>("/api/alerts/extract", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ticker, thesis, action }),
  });
}

export async function createAlerts(
  alerts: Omit<ThesisAlert, "id" | "status" | "triggered_at" | "triggered_price" | "last_price" | "last_checked_at" | "created_at">[]
): Promise<{ created: ThesisAlert[]; demo: boolean }> {
  try {
    const created = await api<ThesisAlert[]>("/api/alerts", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(alerts),
    });
    return { created, demo: false };
  } catch {
    const local = readLocal();
    const created = alerts.map((a, i) => ({
      ...a,
      id: Date.now() + i,
      status: "active" as AlertStatus,
      triggered_at: null,
      triggered_price: null,
      last_price: null,
      last_checked_at: null,
      created_at: new Date().toISOString(),
    }));
    local.created.unshift(...created);
    writeLocal(local);
    return { created, demo: true };
  }
}

export async function dismissAlert(id: number): Promise<{ demo: boolean }> {
  try {
    await api(`/api/alerts/${id}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status: "dismissed" }),
    });
    return { demo: false };
  } catch {
    const local = readLocal();
    if (!local.dismissed.includes(id)) local.dismissed.push(id);
    local.created = local.created.map((a) => (a.id === id ? { ...a, status: "dismissed" as AlertStatus } : a));
    writeLocal(local);
    return { demo: true };
  }
}

export async function scanJournal(): Promise<ScanResult> {
  return api<ScanResult>("/api/alerts/scan-journal", { method: "POST" });
}

/** Distancia del precio actual al nivel, en % firmado (negativo = margen agotándose). */
export function distanceToLevel(alert: ThesisAlert): number | null {
  if (!alert.last_price) return null;
  return ((alert.last_price - alert.level) / alert.level) * 100 * (alert.condition === "price_below" ? 1 : -1);
}

/** Heurística cliente (modo demo): mismos criterios que el backend — regex de niveles explícitos. */
export function clientExtract(thesis: string, action: string): Proposal[] {
  const levels: number[] = [];
  const re = /[$€£]\s?(\d+(?:[.,]\d+)?)|(\d+(?:[.,]\d+)?)\s?[$€£]|\blos\s+(\d+(?:[.,]\d+)?)\b/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(thesis))) {
    const raw = m[1] || m[2] || m[3];
    const v = parseFloat(raw.replace(",", "."));
    if (!isNaN(v)) levels.push(v);
  }
  const bullish = action !== "sell";
  const condition = (bullish ? "price_below" : "price_above") as AlertCondition;
  const uniq = [...new Set(levels)].sort((a, b) => (bullish ? b - a : a - b)).slice(0, 2);
  if (uniq.length === 2) {
    return [
      { condition, level: uniq[0], severity: "aviso", rationale: `Primer nivel citado en la tesis: ${uniq[0]}.` },
      { condition, level: uniq[1], severity: "invalidacion", rationale: `Nivel de ruptura de la tesis: ${uniq[1]}.` },
    ];
  }
  if (uniq.length === 1) {
    return [{ condition, level: uniq[0], severity: "invalidacion", rationale: `Único nivel citado en la tesis: ${uniq[0]}.` }];
  }
  return [];
}
